sklearn_numba_dpex/kmeans/kernels/lloyd_single_step.py

```diff
@@ -249,6 +275,24 @@ def fused_lloyd_single_step(
         # End of outer loop. By now min_idx and min_sample_pseudo_inertia
         # contains the expected values.
 
+        # NB: this check can't be moved at the top at the kernel, because if a work item
+        # exits early with a `return` it will never reach the barriers, thus causing a
+        # deadlock. Early returns are only possible when there are no barriers within
+        # the remaining set of instructions for the running kernel.
+        if sample_idx >= n_samples:
+            return
+
+        if with_sample_weight:
+            weight = sample_weight[sample_idx]
+
+        if return_inertia:
+            inertia = min_sample_pseudo_inertia
+            if compute_exact_inertia:
+                inertia = X_l2_norm + (two * min_sample_pseudo_inertia)
+            if with_sample_weight:
+                inertia = inertia * weight
+            per_sample_inertia[sample_idx] = inertia
```

Comment:
I think that for the name of the variables there's a confusion between "inertia" and "distance to closest centroid" now with the introduction of "sample_weight" in the mix.

But overall I think the logic is good ? "pseudo_inertia" refers to inertia up to a constant that depends on X and is constant across iterations, I think this still holds even with sample_weight ? (that the constant changes does not matter, basically the constant = the weighted sums of squared l2 norms of X samples)

(NB: in practice it's not used anymore by the driver, I let this option in case we want to return to the previous operation where "best inertia" and "best center" are tracked, should I completely remove it ?)
-

Reply:
+1 for cleaning up YAGNIs.

Reply:
:heavy_check_mark: YAGNI cleant up